tar/asm/disassemble.go

```diff
@@ -119,20 +118,27 @@ func NewInputTarStream(r io.Reader, p storage.Packer, fp storage.FilePutter) (io
 			}
 		}
 
-		// it is allowable, and not uncommon that there is further padding on the
-		// end of an archive, apart from the expected 1024 null bytes.
-		remainder, err := ioutil.ReadAll(outputRdr)
-		if err != nil && err != io.EOF {
-			pW.CloseWithError(err)
-			return
-		}
-		_, err = p.AddEntry(storage.Entry{
-			Type:    storage.SegmentType,
-			Payload: remainder,
-		})
-		if err != nil {
-			pW.CloseWithError(err)
-			return
+		// It is allowable, and not uncommon that there is further padding on
+		// the end of an archive, apart from the expected 1024 null bytes. We
+		// do this in chunks rather than in one go to avoid cases where a
+		// maliciously crafted tar file tries to trick us into reading many GBs
+		// into memory.
+		const paddingChunkSize = 1024 * 1024
+		var paddingChunk [paddingChunkSize]byte
+		for {
+			n, err := outputRdr.Read(paddingChunk[:])
+			if err != nil && err != io.EOF {
+				pW.CloseWithError(err)
+				return
+			}
+			_, err = p.AddEntry(storage.Entry{
+				Type:    storage.SegmentType,
+				Payload: paddingChunk[:n],
+			})
+			if err != nil {
+				pW.CloseWithError(err)
+				return
+			}
 		}
 		pW.Close()
```

Comment:
there is no break in the loop, so the pipe writer is never closed.

Reply:
Oops, I will fix that. Working on a test atm.

Reply:
maybe like
```diff
diff --git a/tar/asm/disassemble.go b/tar/asm/disassemble.go
index 0c4edae..5a0d280 100644
--- a/tar/asm/disassemble.go
+++ b/tar/asm/disassemble.go
@@ -125,12 +125,16 @@ func NewInputTarStream(r io.Reader, p storage.Packer, fp storage.FilePutter) (io
                // into memory.
                const paddingChunkSize = 1024 * 1024
                var paddingChunk [paddingChunkSize]byte
+               isEOF := false
                for {
                        n, err := outputRdr.Read(paddingChunk[:])
                        if err != nil && err != io.EOF {
                                pW.CloseWithError(err)
                                return
                        }
+                       if err == io.EOF {
+                               isEOF = true
+                       }
                        _, err = p.AddEntry(storage.Entry{
                                Type:    storage.SegmentType,
                                Payload: paddingChunk[:n],
@@ -139,6 +143,9 @@ func NewInputTarStream(r io.Reader, p storage.Packer, fp storage.FilePutter) (io
                                pW.CloseWithError(err)
                                return
                        }
+                       if isEOF {
+                               break
+                       }
                }
                pW.Close()
        }()
```

Reply:
Do we need to write a null slice in the case of `io.EOF`? Why not skip it entirely?